### PR TITLE
Adding information about when to use `self`

### PR DESCRIPTION
**Pull request by mitulmanish**
- Adds general etiquette about when to use `self` and when to avoid it.

### CODE REVIEW

**Review by sydsrirak — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -47,6 +47,7 @@ The document uses the [raywenderlich.com Swift styleguide](https://github.com/ra
   * [Failing Guards](#failing-guards)
 * [Semicolons](#semicolons)
 * [Parentheses](#parentheses)
+* [Use of self](#use-of-self)
```
Comment (mokagio) [inline README.md]:
```suggestion
* [Use of `self`](#use-of-self)
```

File: README.md [hunk 2]
```diff
@@ -1048,6 +1049,12 @@ if (name == "Hello") {
 }
 ```
 
+## Use of self
```
Comment (mokagio) [inline README.md]:
```suggestion
## Use of `self`
```

File: README.md [hunk 3]
```diff
@@ -1048,6 +1049,12 @@ if (name == "Hello") {
...
+
+Don't use `self`, unless your code can't compile without it.
```
Comment (mokagio) [inline README.md:1054]: Nicely put.

File: README.md [hunk 4]
```diff
@@ -1048,6 +1049,12 @@ if (name == "Hello") {
...
+
+Use `self` only when required by the compiler (in @escaping closures, or in initializers to disambiguate properties from arguments).
```
Comment (mokagio) [inline README.md]:
```suggestion
Use `self` only when required by the compiler (in `@escaping` closures, or in initializers to disambiguate properties from arguments).
```

In the case of the initializers though is worth making an exception:

Consider the following:

```swift
init(price: Int, color: UIColor) {
  self.price = price
  theColor = color
}
```

It doesn't read as homogeneous as the version with `self.` even if it's not necessary to disambiguate.

```swift
init(price: Int, color: UIColor) {
  self.price = price
  self.theColor = color
}
```

The exception to the exception (🙃) would be `init(from decoder:)` as we discussed, where non of the properties are ambiguous so no `self.` is required.

**Review by farhan2106 — APPROVED** [review]
LGTM

**Review by mokagio — APPROVED** [review]

**Pull request merged**